src/Type/TypeCombinator.php

```diff
@@ -636,96 +636,79 @@ private static function processArrayTypes(array $arrayTypes, array $accessoryTyp
 			return [
 				self::intersect(new ArrayType(
 					self::union(...$keyTypesForGeneralArray),
-					self::union(...self::optimizeConstantArrays($valueTypesForGeneralArray)),
+					self::union(...$valueTypesForGeneralArray),
 				), ...$accessoryTypes),
 			];
 		}
 
-		$reducedArrayTypes = self::reduceArrays($arrayTypes);
-
 		return array_map(
 			static fn (Type $arrayType) => self::intersect($arrayType, ...$accessoryTypes),
-			self::optimizeConstantArrays($reducedArrayTypes),
+			self::reduceArrays($arrayTypes),
 		);
 	}
 
 	/**
-	 * @param Type[] $types
+	 * @param Type[] $constantArrays
 	 * @return Type[]
 	 */
-	private static function optimizeConstantArrays(array $types): array
+	private static function reduceArrays(array $constantArrays): array
 	{
-		$constantArrayValuesCount = self::countConstantArrayValueTypes($types);
-
-		if ($constantArrayValuesCount > ConstantArrayTypeBuilder::ARRAY_COUNT_LIMIT) {
-			$results = [];
-			foreach ($types as $type) {
-				$results[] = TypeTraverser::map($type, static function (Type $type, callable $traverse): Type {
-					if ($type instanceof ConstantArrayType) {
-						return $type->generalize(GeneralizePrecision::moreSpecific());
-					}
+		$newArrays = [];
+		$arraysToProcess = [];
+		$emptyArray = null;
+		foreach ($constantArrays as $constantArray) {
+			if (!$constantArray instanceof ConstantArrayType) {
+				$newArrays[] = $constantArray;
+				continue;
+			}
 
-					return $traverse($type);
-				});
+			if ($constantArray->isEmpty()) {
+				$emptyArray = $constantArray;
+				continue;
 			}
 
-			return $results;
+			$arraysToProcess[] = $constantArray;
 		}
 
-		return $types;
-	}
+		if ($emptyArray !== null) {
+			$newArrays[] = $emptyArray;
+		}
 
-	/**
-	 * @param Type[] $types
-	 */
-	private static function countConstantArrayValueTypes(array $types): int
-	{
-		$constantArrayValuesCount = 0;
-		foreach ($types as $type) {
-			if ($type instanceof ConstantArrayType) {
-				$constantArrayValuesCount += count($type->getValueTypes());
+		$arraysToProcessPerKey = [];
+		foreach ($arraysToProcess as $i => $arrayToProcess) {
+			foreach ($arrayToProcess->getKeyTypes() as $keyType) {
+				$arraysToProcessPerKey[$keyType->getValue()][] = $i;
 			}
+		}
 
-			TypeTraverser::map($type, static function (Type $type, callable $traverse) use (&$constantArrayValuesCount): Type {
-				if ($type instanceof ConstantArrayType) {
-					$constantArrayValuesCount += count($type->getValueTypes());
-				}
+		$eligibleCombinations = [];
 
-				return $traverse($type);
-			});
+		foreach ($arraysToProcessPerKey as $arrays) {
+			for ($i = 0, $arraysCount = count($arrays); $i < $arraysCount - 1; $i++) {
+				for ($j = $i + 1; $j < $arraysCount; $j++) {
+					$eligibleCombinations[$arrays[$i]][$arrays[$j]] = $arrays[$j];
+				}
+			}
 		}
```

Comment:
This is the main bit - it reduces the combinations that will be inspected to only combinations that have at least one key in common.